### PR TITLE
Client Travel

### PR DESCRIPTION
**Pull request by joshuahuburn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Client travel working.

Still some small refactoring to be done but please review the overall logic.

Corresponding engine change https://github.com/improbableio/UnrealEngine/pull/24

#### Tests
Made a gif of client travel working.
#### Primary reviewers
@davedolben @Vatyx

### CODE REVIEW

File: SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp [hunk 1]
```diff
@@ -72,9 +97,8 @@ bool USpatialGameInstance::StartGameInstance_SpatialGDKClient(FString& Error)
 		GetEngine()->ShutdownWorldNetDriver(GetWorldContext()->World());
 	}
 
-	// This will use the URL / Map that was setup by the server worker (which is loaded first).
-	// By not specifying a hostname the connection defaults to local.
 	FURL URL = WorldContext->LastURL;
+	URL.Host = "127.0.0.1";
```
Comment (joshuahuburn) [inline SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp]: Needed to prevent the host from being empty and triggering the wrong type of travel in a PIE context.

File: SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp [hunk 2]
```diff
@@ -51,9 +51,34 @@ bool USpatialGameInstance::HasSpatialNetDriver() const
 	return bHasSpatialNetDriver;
 }
 
+void MySpatialBrowse(FWorldContext& WorldContext, FURL URL)
```
Comment (Vatyx) [inline SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp]: I think this should be renamed?

File: SpatialGDK/Source/Private/Interop/SpatialPlayerSpawner.cpp [hunk 3]
```diff
@@ -56,7 +56,7 @@ void USpatialPlayerSpawner::SendPlayerSpawnRequest()
 
 void USpatialPlayerSpawner::ReceivePlayerSpawnResponse(Worker_CommandResponseOp& Op)
 {
-	UE_LOG(LogTemp, Warning, TEXT("Got a response"));
+	UE_LOG(LogTemp, Warning, TEXT("ReceivePlayerSpawnResponse"));
```
Comment (Vatyx) [inline SpatialGDK/Source/Private/Interop/SpatialPlayerSpawner.cpp]: I'm deleting this log since it was pretty useless in my other PR.

**Comment by davedolben** [comment]
Your branch name is funky. There's an extra `origin/` on there

File: SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp [hunk 4]
```diff
@@ -51,9 +51,34 @@ bool USpatialGameInstance::HasSpatialNetDriver() const
...
+		WorldContext.PendingNetGame = NewObject<USpatialPendingNetGame>();
+		WorldContext.PendingNetGame->Initialize(URL);
+		// See above comment about setting NetDriver manually here.
+		WorldContext.PendingNetGame->NetDriver = TempNetDriver;
+	}
+}
 
 bool USpatialGameInstance::StartGameInstance_SpatialGDKClient(FString& Error)
 {
+	GetEngine()->SpatialBrowseDelegate = &MySpatialBrowse;
```
Comment (davedolben) [inline SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp]: Do we know that this is early enough to set the delegate? Is it possible that Browse will have already been called at this point? My guess is no, but it would be good to verify.

File: SpatialGDK/Source/Private/EngineClasses/SpatialPackageMapClient.cpp [hunk 5]
```diff
@@ -202,21 +226,31 @@ void FSpatialNetGUIDCache::RemoveEntitySubobjectsNetGUIDs(Worker_EntityId Entity
...
+	// If we have paths, network-sanitize all of them (e.g. removing PIE prefix).
+	if (ObjectRef.Path.IsSet())
 	{
-		FNetworkGUID OuterGUID;
-		if (ObjectRef.Outer.IsSet())
-		{
-			OuterGUID = GetNetGUIDFromUnrealObjectRef(ObjectRef.Outer.GetValue());
+		improbable::UnrealObjectRef* Iterator = &ObjectRef;
+		while (true) {
+			if (!Iterator->Path.IsSet())
```
Comment (davedolben) [inline SpatialGDK/Source/Private/EngineClasses/SpatialPackageMapClient.cpp]: Invert this if statement and remove the commented out log line.

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp [hunk 6]
```diff
@@ -89,37 +89,63 @@ void USpatialNetDriver::OnMapLoaded(UWorld* LoadedWorld)
...
 	if (Connection != nullptr)
 	{
 		return;
 	}
 
 	Connection = NewObject<USpatialWorkerConnection>();
+
+	bool bUseLocator = LoadedWorld->URL.HasOption(TEXT("spatialLocator"));
+
+	if (bUseLocator)
```
Comment (mpmisko) [inline SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: You can directly call `LoadedWorld->URL.HasOption(TEXT("spatialLocator"))`

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp [hunk 7]
```diff
@@ -89,37 +89,63 @@ void USpatialNetDriver::OnMapLoaded(UWorld* LoadedWorld)
...
+	{
+		Connection->LocatorConfig.ProjectName = LoadedWorld->URL.GetOption(TEXT("project"), TEXT(""));
```
Comment (davedolben) [inline SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: I've just realized that this doesn't allow for specifying the deployment name. Looks like the GDK just defaults to the first active deployment in a project, which isn't desired behavior. I'm going to dig around and see if there's a quick solution to this.

File: SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp (hunk 2 above)
Comment (davedolben) [reply SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp]: Done

File: SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp (hunk 4 above)
Comment (davedolben) [reply SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp]: It would appear that the answer is yes.

File: SpatialGDK/Source/Private/EngineClasses/SpatialPackageMapClient.cpp (hunk 5 above)
Comment (davedolben) [reply SpatialGDK/Source/Private/EngineClasses/SpatialPackageMapClient.cpp]: Done

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp (hunk 6 above)
Comment (davedolben) [reply SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Done

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp (hunk 7 above)
Comment (davedolben) [reply SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Added in a fix, although I haven't tested it with a project with multiple running deployments yet

**Comment by davedolben** [comment]
Resolved a bunch of the comments, and tested this with connecting via the locator, so I guess it would be weird for me to approve this PR now...

File: SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp [hunk 8]
```diff
@@ -51,9 +51,34 @@ bool USpatialGameInstance::HasSpatialNetDriver() const
...
+{
+	bool bIsSpatial = false;
+	UNetDriver* TempNetDriver = nullptr;
+	if (GEngine->CreateNamedNetDriver(WorldContext.PendingNetGame, NAME_PendingNetDriver, NAME_GameNetDriver))
+	{
+		TempNetDriver = GEngine->FindNamedNetDriver(WorldContext.PendingNetGame, NAME_PendingNetDriver);
+		// Setting the PendingNetGame's NetDriver is necessary here because the call to CreateNamedNetDriver above will interfere
+		// with the internals of InitNetDriver and cause the NetDriver not to be initialized, and fail a check().
+		WorldContext.PendingNetGame->NetDriver = TempNetDriver;
+		bIsSpatial = TempNetDriver->IsA(USpatialNetDriver::StaticClass());
```
Comment (Vatyx) [inline SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp]: You should use `GetDefault<UGeneralProjectSettings>()->bSpatialNetworking;` instead.

File: SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp (hunk 1 above)
Comment (Vatyx) [inline SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp]: If Host is a `FString` you should do `TEXT()`

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp [hunk 9]
```diff
@@ -89,37 +89,62 @@ void USpatialNetDriver::OnMapLoaded(UWorld* LoadedWorld)
...
+	}
+	else if (LoadedWorld->URL.HasOption(TEXT("spatialReceptionist")))
+	{
+		// Check for overrides in the travel URL.
+		if (!LoadedWorld->URL.Host.IsEmpty())
+		{
+			Connection->ReceptionistConfig.ReceptionistHost = LoadedWorld->URL.Host;
+			Connection->ReceptionistConfig.ReceptionistPort = LoadedWorld->URL.Port;
+		}
+		// Check if we need to use external IPs (if not connecting to localhost).
```
Comment (Vatyx) [inline SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Obvious comment, can be removed.

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp [hunk 10]
```diff
@@ -89,37 +89,62 @@ void USpatialNetDriver::OnMapLoaded(UWorld* LoadedWorld)
...
+		if (Connection->ReceptionistConfig.ReceptionistHost.Compare(TEXT("127.0.0.1")) == 0)
```
Comment (Vatyx) [inline SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Can this IP be moved to SpatialConstants.h?

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp [hunk 11]
```diff
@@ -932,7 +957,7 @@ USpatialNetConnection* USpatialNetDriver::AcceptNewPlayer(const FURL& InUrl, boo
 
 	if (!ErrorMsg.IsEmpty())
 	{
-		UE_LOG(LogNet, Log, TEXT("PreLogin failure: %s"), *ErrorMsg);
+		UE_LOG(LogNet, Error, TEXT("PreLogin failure: %s"), *ErrorMsg);
```
Comment (Vatyx) [inline SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Change to `LogSpatialNetDriver` category.

File: SpatialGDK/Source/Private/EngineClasses/SpatialPackageMapClient.cpp [hunk 12]
```diff
@@ -98,6 +98,30 @@ FNetworkGUID USpatialPackageMapClient::GetNetGUIDFromUnrealObjectRef(const impro
 	return SpatialGuidCache->GetNetGUIDFromUnrealObjectRef(ObjectRef);
 }
 
+FNetworkGUID FSpatialNetGUIDCache::GetNetGUIDFromUnrealObjectRef(const improbable::UnrealObjectRef& ObjectRef)
+{
+	improbable::UnrealObjectRef NetRemappedObjectRef = ObjectRef;
```
Comment (Vatyx) [inline SpatialGDK/Source/Private/EngineClasses/SpatialPackageMapClient.cpp]: This will break once Josh merge's his namespace change so just be aware.

File: SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp (hunk 8 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp]: Done, cheers :D

File: SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp (hunk 1 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp]: Done.

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp (hunk 9 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Done.

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp (hunk 10 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Sure, why not.

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp (hunk 11 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Done.

File: SpatialGDK/Source/Private/Interop/SpatialPlayerSpawner.cpp (hunk 3 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/Private/Interop/SpatialPlayerSpawner.cpp]: Will delete too.

File: SpatialGDK/Source/SpatialGDK.Build.cs [hunk 13]
```diff
@@ -45,7 +45,7 @@ public SpatialGDK(ReadOnlyTargetRules Target) : base(Target)
 			PublicDependencyModuleNames.Add("UnrealEd");
 		}
 
-   		var CoreSdkLibraryDir = Path.GetFullPath(Path.Combine(ModuleDirectory, "..", "Binaries", "ThirdParty", "Improbable", Target.Platform.ToString()));
+		var CoreSdkLibraryDir = Path.GetFullPath(Path.Combine(ModuleDirectory, "..", "Binaries", "ThirdParty", "Improbable", Target.Platform.ToString()));
```
[the remark this answers is not in the document]
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK.Build.cs]: Auto-formatting mess up. Removing.

File: SpatialGDK/Source/Public/Interop/Connection/ConnectionConfig.h [hunk 14]
```diff
@@ -56,11 +56,13 @@ struct FLocatorConfig : public FConnectionConfig
 		const TCHAR* CommandLine = FCommandLine::Get();
 
 		FParse::Value(CommandLine, TEXT("projectName"), ProjectName);
+		FParse::Value(CommandLine, TEXT("deploymentName"), DeploymentName);
```
[the remark this answers is not in the document]
Comment (joshuahuburn) [reply SpatialGDK/Source/Public/Interop/Connection/ConnectionConfig.h]: This is for the `FLocatorConfig` so I believe it's relevant no?

File: SpatialGDK/Source/Private/EngineClasses/SpatialPackageMapClient.cpp [hunk 15]
```diff
@@ -219,6 +226,28 @@ FNetworkGUID FSpatialNetGUIDCache::GetNetGUIDFromUnrealObjectRef(const UnrealObj
 	return NetGUID;
 }
 
+void FSpatialNetGUIDCache::NetworkRemapObjectRefPaths(UnrealObjectRef& ObjectRef) const
+{
+	// If we have paths, network-sanitize all of them (e.g. removing PIE prefix).
+	if (ObjectRef.Path.IsSet())
+	{
+		UnrealObjectRef* Iterator = &ObjectRef;
+		while (true) {
```
Comment (girayimprobable) [inline SpatialGDK/Source/Private/EngineClasses/SpatialPackageMapClient.cpp]: Nit: `{` next line

File: SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp [hunk 16]
```diff
@@ -51,9 +51,36 @@ bool USpatialGameInstance::HasSpatialNetDriver() const
...
+{
+	bool bIsSpatial = false;
+	UNetDriver* TempNetDriver = nullptr;
+	if (GEngine->CreateNamedNetDriver(WorldContext.PendingNetGame, NAME_PendingNetDriver, NAME_GameNetDriver))
+	{
+		TempNetDriver = GEngine->FindNamedNetDriver(WorldContext.PendingNetGame, NAME_PendingNetDriver);
+		// Setting the PendingNetGame's NetDriver is necessary here because the call to CreateNamedNetDriver above will interfere
+		// with the internals of InitNetDriver and cause the NetDriver not to be initialized, and fail a check().
+		WorldContext.PendingNetGame->NetDriver = TempNetDriver;
+		bIsSpatial = GetDefault<UGeneralProjectSettings>()->bSpatialNetworking;
```
Comment (davedolben) [inline SpatialGDK/Source/Private/EngineClasses/SpatialGameInstance.cpp]: This line removes the necessity for this entire if block. The only reason I instantiated the net driver here was to check if it was a `USpatialNetDriver` and set the value of `bIsSpatial`. I think you can now refactor this function into just a few lines:

```
if (GetDefault<UGeneralProjectSettings>()->bSpatialNetworking)
{
  WorldContext.PendingNetGame = NewObject<USpatialPendingNetGame>();
  WorldContext.PendingNetGame->Initialize(URL);
}
```

File: SpatialGDK/Source/Private/Interop/Connection/SpatialWorkerConnection.cpp [hunk 17]
```diff
@@ -3,6 +3,7 @@
 #include "SpatialWorkerConnection.h"
 
 #include "Async/Async.h"
+#include <string>
```
Comment (davedolben) [inline SpatialGDK/Source/Private/Interop/Connection/SpatialWorkerConnection.cpp]: Don't need this any more. Sorry, added it in an experiment and forgot to remove it.

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp [hunk 18]
```diff
@@ -89,28 +89,55 @@ void USpatialNetDriver::OnMapLoaded(UWorld* LoadedWorld)
...
+	UE_LOG(LogSpatialOSNetDriver, Log, TEXT("Loaded Map %s. Connecting to SpatialOS."), *LoadedWorld->GetName());
+
 	if (Connection != nullptr)
 	{
 		return;
 	}
 
 	Connection = NewObject<USpatialWorkerConnection>();
+
+	if (LoadedWorld->URL.HasOption(TEXT("spatialLocator")))
```
Comment (girayimprobable) [inline SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Awesome!

File: SpatialGDK/Source/Public/Interop/Connection/ConnectionConfig.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (Vatyx) [reply SpatialGDK/Source/Public/Interop/Connection/ConnectionConfig.h]: If it's only relevant for `FLocatorConfig` then it should be in that class.

But this isn't needed because we dynamically grab the deployment list. Look at `USpatialWorkerConnection::ConnectToLocator()`

File: SpatialGDK/Source/Public/SpatialConstants.h [hunk 19]
```diff
@@ -36,4 +36,6 @@ namespace SpatialConstants
 		uint32 WaitTimeExponentialFactor = 1u << (NumAttempts - 1);
 		return FIRST_COMMAND_RETRY_WAIT_SECONDS * WaitTimeExponentialFactor;
 	}
+
+	const FString LOCAL_HOST = TEXT("127.0.0.1");
```
Comment (Vatyx) [inline SpatialGDK/Source/Public/SpatialConstants.h]: Super nit: No underscore, since you usually type `localhost`.

File: SpatialGDK/Source/Public/Interop/Connection/ConnectionConfig.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (davedolben) [reply SpatialGDK/Source/Public/Interop/Connection/ConnectionConfig.h]: Yup, this line should be in the constructor for `FLocatorConfig` (I think the original line is lost b/c of a force push...)

Also `ConnectToLocator()` just connects to the first deployment that you get from the list. The point of this flag was to search that list for a deployment with the corresponding name and connect to it specifically, otherwise to default to the first deployment in the project.

File: SpatialGDK/Source/Private/Interop/SpatialPlayerSpawner.cpp [hunk 20]
```diff
@@ -56,8 +56,6 @@ void USpatialPlayerSpawner::SendPlayerSpawnRequest()
 
 void USpatialPlayerSpawner::ReceivePlayerSpawnResponse(Worker_CommandResponseOp& Op)
 {
-	UE_LOG(LogTemp, Warning, TEXT("Got a response"));
```
Comment (danielimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialPlayerSpawner.cpp]: Driveby fix?

File: SpatialGDK/Source/Public/Interop/Connection/SpatialWorkerConnection.h [hunk 21]
```diff
@@ -36,6 +36,9 @@ class SPATIALGDK_API USpatialWorkerConnection : public UObject
 
 	OnConnectedDelegate OnConnected;
 
+	FReceptionistConfig ReceptionistConfig;
```
Comment (danielimprobable) [inline SpatialGDK/Source/Public/Interop/Connection/SpatialWorkerConnection.h]: Could we use accessors so that is easier to track where they are changed?

**Comment by davedissian** [comment]
Just an FYI. The locator flow is eventually getting replaced with "locator v2" (that removes the need for queuing). The changes to the C API in 14.0 are here: https://github.com/improbable/everything/pull/2615/files#diff-b33ca74dd23bc5964c77f9922084210a

Essentially. the flow becomes the following:
- You pass your Xbox Live/Steam/PSN credentials to a login service managed by the game developer / improbably via gRPC (undecided). You get back a "player identity token"
- You then pass the "player identity token" to our login service and you get a "login token" for each deployment, potentially after some queuing depending on the implementation.
- You pass the PIT and LT pair to the locator, and immediately get a worker connection (as the queuing is done in the previous "login token" step).

Essentially, the GDK would eventually need to support passing these two tokens to the worker SDK when connecting via that flow.

Doesn't need to be supported in this PR, but its worth keeping it in mind as this is the login flow that Midwinter will probably end up using. I'll probably implement it myself when working on the Xbox certification demo. Maybe it's worth calling the locator flow "legacyLocator" or something.

File: SpatialGDK/Source/Public/SpatialConstants.h (hunk 19 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/Public/SpatialConstants.h]: Voted here, underscore won <3

File: SpatialGDK/Source/Public/Interop/Connection/SpatialWorkerConnection.h (hunk 21 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/Public/Interop/Connection/SpatialWorkerConnection.h]: Discussed offline, leaving as public for now.

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp [hunk 22]
```diff
@@ -89,28 +89,55 @@ void USpatialNetDriver::OnMapLoaded(UWorld* LoadedWorld)
 		return;
 	}
 
-	UE_LOG(LogSpatialOSNetDriver, Log, TEXT("Loaded Map %s. Connecting to SpatialOS."), *LoadedWorld->GetName());
-
 	//checkf(!SpatialOSInstance->IsConnected(), TEXT("SpatialOS should not be connected already. This is probably because we attempted to travel to a different level, which current isn't supported."));
```
[the remark this answers is not in the document]
Comment (joshuahuburn) [reply SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Believe this is being handled by DTB cleanup PR #3 (since this variable no longer exists) so don't want to hit it too :D

File: SpatialGDK/Source/Public/Interop/Connection/ConnectionConfig.h [hunk 23]
```diff
@@ -35,7 +36,7 @@ struct FConnectionConfig
 struct FReceptionistConfig : public FConnectionConfig
 {
 	FReceptionistConfig()
-		: ReceptionistHost(TEXT("127.0.0.1"))
+		: ReceptionistHost(SpatialConstants::LOCAL_HOST)
 		, ReceptionistPort(7777)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/Public/Interop/Connection/ConnectionConfig.h]: Nit: make default port a `SpatialConstant` as well?

Comment (joshuahuburn) [reply SpatialGDK/Source/Public/Interop/Connection/ConnectionConfig.h]: Sure.
